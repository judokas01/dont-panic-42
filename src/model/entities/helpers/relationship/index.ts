/**
 * These relation helpers are here to provide maximum flexibility for the storage medium of the application entities.
 * The relations are possibly "not loaded" at all times and will require the storage medium to provide the relation data if used.
 *
 * A runtime error will be thrown if the relation data are not available when accessed.
 * This will prompt the developer to make sure the data is available via the storage repository.
 */

import { RelationNotLoadedError } from '@root/model/entities/helpers/relationship/errors'

type Id = string | number

export type EntityLike<T> = {
    id?: Id
    clone(): T
}

export class HasOne<T extends EntityLike<T>> {
    private constructor(
        private readonly entity: T | null,
        private readonly errPath: string,
        private readonly id?: T['id'],
    ) {}

    static unloaded<T extends EntityLike<T>>(path: string, id?: T['id']) {
        return new HasOne<T>(null, path, id)
    }

    static loaded<T extends EntityLike<T>>(path: string, entity: T) {
        return new HasOne<T>(entity, path, entity.id)
    }

    hasId = () => Boolean(this.id)

    isLoaded = () => Boolean(this.entity)

    getId = (): T['id'] => {
        if (!this.id) throw new RelationNotLoadedError(this.errPath)

        return this.id
    }

    get = (): Readonly<T> => {
        if (!this.isLoaded()) throw new RelationNotLoadedError(this.errPath)

        return this.entity as T
    }
}

export class HasMany<T extends EntityLike<T>> {
    private constructor(
        private readonly entities: Readonly<T>[] | null,
        private readonly path: string,
    ) {}

    /**
     * @deprecated use `HasMany.loaded([], path)` instead
     */
    static empty(path: string) {
        return new HasMany<never>([], path)
    }

    static unloaded(path: string) {
        return new HasMany<never>(null, path)
    }

    static loaded<T extends EntityLike<T>>(path: string, entities: T[]) {
        return new HasMany<T>(entities, path)
    }

    isLoaded = () => Boolean(this.entities)

    get = (): Readonly<Readonly<T>[]> => {
        if (!this.isLoaded()) throw new RelationNotLoadedError(this.path)

        return this.entities as T[]
    }
}
