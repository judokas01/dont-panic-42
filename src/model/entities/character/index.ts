import { HasMany } from '../helpers/relationship'
import { Clonable } from '../helpers/relationship/clone'
import type { Nemesis } from '../nemesis'
import { Immutable } from '@root/model/lib/typescript'

export class Character implements Clonable<Character> {
    constructor(private character: CharacterData) {}

    get id(): Immutable<CharacterData['id']> {
        return this.character.id
    }

    get data(): Immutable<CharacterData> {
        return this.character
    }

    get age(): number {
        return this.diffInYears(new Date(), this.character.born)
    }

    get yearsInSpace(): number {
        return this.diffInYears(new Date(), this.character.inSpaceSince ?? new Date())
    }

    private diffInYears = (date1: Date, date2: Date): number => {
        const differenceInMilliseconds = Math.abs(date1.getTime() - date2.getTime())
        const differenceInYears = Math.floor(differenceInMilliseconds / (1000 * 60 * 60 * 24 * 365))
        return differenceInYears
    }

    clone = (): Character => new Character({ ...this.character })
}

export type CharacterData = {
    id: number
    name: string
    gender?: string
    ability: string
    minimalDistance: number
    weight: number
    born: Date
    inSpaceSince: Date | null
    beerConsumption: number
    knowsTheAnswer: boolean
    /**
     * SelfNote: plural of nemesis is `nemeses`
     */
    nemeses: HasMany<Nemesis>
}
