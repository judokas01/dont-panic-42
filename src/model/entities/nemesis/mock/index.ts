import { faker } from '@faker-js/faker'
import { CreateNemesisData } from '..'
import { HasOne } from '../../helpers/relationship'
import { coinFlip } from '@root/model/test-utils/tools'

export const getRandomNemesisData = ({
    character = HasOne.unloaded('nemesis.character'),
    isAlive = coinFlip(),
    years = faker.number.int({ min: 0, max: 10000 }),
}: Partial<CreateNemesisData> = {}): CreateNemesisData => ({
    character,
    isAlive,
    years,
})
