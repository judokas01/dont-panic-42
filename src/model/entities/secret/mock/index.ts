import { faker } from '@faker-js/faker'
import { CreateSecretData } from '..'
import { HasOne } from '../../helpers/relationship'

export const getRandomSecretData = ({
    nemesis = HasOne.unloaded('secret.nemesis'),
    secretCode = faker.number.bigInt(),
}: Partial<CreateSecretData> = {}): CreateSecretData => ({
    nemesis,
    secretCode,
})
