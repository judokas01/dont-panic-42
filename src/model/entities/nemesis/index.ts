import { Character } from '../character'
import { HasMany, HasOne } from '../helpers/relationship'
import { Clonable } from '../helpers/relationship/clone'
import type { Secret } from '../secret'
import { Immutable } from '@root/model/lib/typescript'

export class Nemesis implements Clonable<Nemesis> {
    constructor(private nemesis: NemesisData) {}

    get id(): Immutable<NemesisData['id']> {
        return this.nemesis.id
    }

    get data(): Immutable<NemesisData> {
        return this.nemesis
    }

    clone = (): Nemesis => new Nemesis({ ...this.nemesis })
}

export type NemesisData = {
    id: number
    isAlive: boolean
    years?: number
    character: HasOne<Character>
    secrets: HasMany<Secret>
}

export type CreateNemesisData = Omit<NemesisData, 'id' | 'secrets'>
