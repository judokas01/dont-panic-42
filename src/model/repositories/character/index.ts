import { Injectable } from '@nestjs/common'
import { toCoreCharacter } from '../common/mappers'
import { PrismaService } from '@root/infrastructure/prisma/client'
import { Character, CreateCharacterData } from '@root/model/entities/character'

@Injectable()
export class CharacterRepository {
    constructor(private prisma: PrismaService) {}

    insertOne = async (data: CreateCharacterData): Promise<Character> => {
        const created = await this.prisma.character.create({
            data: {
                ability: data.ability,
                beer_consumption: data.beerConsumption,
                born: data.born,
                knows_the_answer: data.knowsTheAnswer,
                minimal_distance: data.minimalDistance,
                name: data.name,
                gender: data.gender,
                in_space_since: data.inSpaceSince,
                weight: data.weight,
            },
        })

        return toCoreCharacter(created)
    }

    findAll = async (): Promise<Character[]> => {
        const found = await this.prisma.character.findMany({
            include: {
                nemeses: {
                    include: {
                        secrets: true,
                    },
                },
            },
        })

        return found.map(toCoreCharacter)
    }

    clear = async (): Promise<void> => {
        await this.prisma.character.deleteMany()
    }
}
