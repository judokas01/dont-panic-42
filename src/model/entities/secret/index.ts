import { HasOne } from '../helpers/relationship'
import { Clonable } from '../helpers/relationship/clone'
import { Nemesis } from '../nemesis'
import { Immutable } from '@root/model/lib/typescript'

export class Secret implements Clonable<Secret> {
    constructor(private secret: SecretData) {}

    get id(): Immutable<SecretData['id']> {
        return this.secret.id
    }

    get data(): Immutable<SecretData> {
        return this.secret
    }

    clone = (): Secret => new Secret({ ...this.secret })
}

export type SecretData = {
    id: number
    secretCode: bigint
    nemesis: HasOne<Nemesis>
}

export type CreateSecretData = Omit<SecretData, 'id'>
