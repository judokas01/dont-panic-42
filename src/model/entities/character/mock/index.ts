import { faker } from '@faker-js/faker'
import { CreateCharacterData } from '..'
import { getRandomNemesisData } from '../../nemesis/mock'
import { HasOne } from '../../helpers/relationship'
import { getRandomSecretData } from '../../secret/mock'
import { NemesisData } from '../../nemesis'
import { CreateSecretData } from '../../secret'
import { TestingAppRepositories } from '@root/model/test-utils/module'
import { coinFlip } from '@root/model/test-utils/tools'

export const getRandomCharacterData = ({
    ability = 'ability',
    beerConsumption = faker.number.int({ min: 0, max: 10000 }),
    born = faker.date.past(),
    gender = faker.person.gender(),
    inSpaceSince = coinFlip() ? faker.date.past() : null,
    knowsTheAnswer = coinFlip(),
    minimalDistance = faker.number.int({ min: 0, max: 10000 }),
    name = faker.person.fullName(),
    weight = faker.number.int({ min: 0, max: 10000 }),
}: Partial<CreateCharacterData> = {}): CreateCharacterData => ({
    ability,
    beerConsumption,
    born,
    inSpaceSince,
    knowsTheAnswer,
    minimalDistance,
    name,
    weight,
    gender,
})

export const createCharacterSet = async (
    overrides: {
        character?: Partial<CreateCharacterData>
        nemeses?: {
            count?: number
            override?: (i?: number) => Partial<NemesisOverride>
        }
    },
    repositories: TestingAppRepositories,
) => {
    const { character, nemeses } = overrides
    const createdCharacter = await repositories.characterRepository.insertOne(
        getRandomCharacterData({ ...character }),
    )

    await Promise.all(
        Array.from({ length: nemeses?.count ?? randomCount() }).map(async (_, i) => {
            const nemesis = await repositories.nemesisRepository.insertOne(
                getRandomNemesisData({
                    character: HasOne.loaded('nemesis.character', createdCharacter),
                    ...nemeses?.override?.(i),
                }),
            )

            await Promise.all(
                Array.from({ length: nemeses?.override?.(i)?.secretCount ?? randomCount() }).map(
                    (_, i) =>
                        repositories.secretRepository.insertOne(
                            getRandomSecretData({
                                nemesis: HasOne.loaded('secret.nemesis', nemesis),
                                ...nemeses?.override?.(i)?.secretOverride?.(i),
                            }),
                        ),
                ),
            )
        }),
    )
}

type NemesisOverride = Omit<NemesisData, 'id' | 'character'> & {
    secretCount?: number
    secretOverride?: (i?: number) => Partial<CreateSecretData>
}

export const randomCount = () => faker.number.int({ min: 0, max: 5 })
