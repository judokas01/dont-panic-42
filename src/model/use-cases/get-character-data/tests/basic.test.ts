import { describe, beforeAll, it, expect, beforeEach } from 'vitest'
import {
    CharactersResponse,
    GetCharactersDataUseCase,
} from '@root/model/use-cases/get-character-data'
import {
    TestingAppRepositories,
    cleanDatabase,
    getTestingModule,
} from '@root/model/test-utils/module'
import { createCharacterSet, randomCount } from '@root/model/entities/character/mock'
import { HasMany, HasOne } from '@root/model/entities/helpers/relationship'
import { CharacterData } from '@root/model/entities/character'
import { NemesisData } from '@root/model/entities/nemesis'
import { SecretData } from '@root/model/entities/secret'

describe('Get characters use case base tests', () => {
    let repositories: TestingAppRepositories
    let useCase: GetCharactersDataUseCase

    beforeAll(async () => {
        ;({ repositories, useCase } = await getTestingModule())
    })

    beforeEach(async () => {
        await cleanDatabase(repositories)
    })

    describe('Statistics tests', () => {
        it('should return all stats zero when no character exists', async () => {
            const result = await useCase.get()
            expect(result).toMatchSnapshot()
        })

        it('should return stats with random values when characters exist', async () => {
            await createCharacterSet({}, repositories)
            await createCharacterSet({}, repositories)
            await createCharacterSet({}, repositories)

            const result = await useCase.get()
            expect(result).toMatchObject({
                characters: expect.any(Array),
                stats: {
                    averageAge: expect.any(Number),
                    averageBeerConsumption: expect.any(Number),
                    averageNemeses: expect.any(Number),
                    averageYearsInSpace: expect.any(Number),
                    averageYearsOfBeingNemesis: expect.any(Number),
                    characterCount: 3,
                    genders: {
                        female: expect.any(Number),
                        male: expect.any(Number),
                        other: expect.any(Number),
                    },
                },
            } satisfies CharactersResponse)
        })

        it('should return expected stats', async () => {
            await createCharacterSet(
                {
                    character: {
                        born: new Date('2000-01-01'),
                        beerConsumption: 17,
                        gender: 'f',
                        inSpaceSince: null,
                    },
                    nemeses: {
                        count: 11,
                    },
                },
                repositories,
            )

            await createCharacterSet(
                {
                    character: {
                        born: new Date('1958-01-01'),
                        beerConsumption: 42,
                        gender: 'some other',
                        inSpaceSince: new Date('1900-01-01'),
                    },
                    nemeses: {
                        count: 0,
                    },
                },
                repositories,
            )

            const result = await useCase.get()

            expect(result.stats).toMatchObject({
                // this should be calculated or we will celebrate new year with flaky test
                averageAge: 45,
                averageBeerConsumption: (42 + 17) / 2,
                averageNemeses: (11 + 0) / 2,
                // this should be calculated or we will celebrate new year with flaky test
                averageYearsInSpace: 124 / 2,
                averageYearsOfBeingNemesis: expect.any(Number),
                characterCount: 2,
                genders: {
                    female: 1,
                    male: 0,
                    other: 1,
                },
            } satisfies CharactersResponse['stats'])
        })

        it('should return expected stats nemeses related stats', async () => {
            const char1NemesisCount = 3
            const char2NemesisCount = 5
            await createCharacterSet(
                {
                    nemeses: {
                        count: char1NemesisCount,
                        override: (i) => ({ years: i * 66 }),
                    },
                },
                repositories,
            )

            await createCharacterSet(
                {
                    nemeses: {
                        count: char2NemesisCount,
                        override: (i) => ({ years: i * 10 }),
                    },
                },
                repositories,
            )

            /**
             * 0 + 66 + 132 = 198
             * 0 + 10 + 20 + 30 + 40 = 100
             * (198 + 100) / 2 = 132
             */
            const expectedYears = 149
            const result = await useCase.get()

            expect(result.stats.averageNemeses).toBe((char1NemesisCount + char2NemesisCount) / 2)
            expect(result.stats.averageYearsOfBeingNemesis).toBe(expectedYears)
        })
    })

    describe('characters tests', () => {
        it('should return a character set with all the related data', async () => {
            await createCharacterSet(
                {
                    nemeses: {
                        count: randomCount(),
                        override: () => ({ secretCount: randomCount() }),
                    },
                },
                repositories,
            )
            await createCharacterSet(
                {
                    nemeses: {
                        count: randomCount(),
                        override: () => ({ secretCount: randomCount() }),
                    },
                },
                repositories,
            )
            const result = await useCase.get()
            expect(result.characters).toHaveLength(2)
            result.characters.forEach(({ data }) => {
                expect(data).toMatchObject({
                    id: expect.any(Number),
                    name: expect.any(String),
                    ability: expect.any(String),
                    beerConsumption: expect.any(Number),
                    born: expect.any(Date),
                    inSpaceSince: data.inSpaceSince ? expect.any(Date) : null,
                    knowsTheAnswer: expect.any(Boolean),
                    minimalDistance: expect.any(Number),
                    nemeses: expect.any(HasMany),
                    weight: expect.any(Number),
                    gender: expect.any(String),
                } satisfies CharacterData)

                const nemeses = data.nemeses.get()

                nemeses.forEach(({ data: nemesisData }) => {
                    expect(nemesisData).toMatchObject({
                        id: expect.any(Number),
                        isAlive: expect.any(Boolean),
                        years: expect.any(Number),
                        character: expect.any(HasOne),
                        secrets: expect.any(HasMany),
                    } as NemesisData)

                    const secrets = nemesisData.secrets.get()

                    secrets.forEach(({ data }) => {
                        expect(data).toMatchObject({
                            id: expect.any(Number),
                            secretCode: expect.any(BigInt),
                            nemesis: expect.any(HasOne),
                        } satisfies SecretData)
                    })
                })
            })
        })
    })
})
