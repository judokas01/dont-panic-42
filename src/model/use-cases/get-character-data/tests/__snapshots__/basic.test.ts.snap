// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Get characters use case base tests > Statistics tests > should return all stats zero when no character exists 1`] = `
{
  "characters": [],
  "stats": {
    "averageAge": 0,
    "averageBeerConsumption": 0,
    "averageNemeses": 0,
    "averageYearsInSpace": 0,
    "averageYearsOfBeingNemesis": 0,
    "characterCount": 0,
    "genders": {
      "female": 0,
      "male": 0,
      "other": 0,
    },
  },
}
`;
